README.md
# NAME

App::CustomOutput - exec command with custom output

# INSTALL

    > curl -LO https://raw.githubusercontent.com/shoichikaji/custom-output/master/bin/custom-output
    > chmod +x custom-output

# SYNOPSIS

    > custom-output --append-time -- perl -v
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] This is perl 5, version 20, subversion 0 (v5.20.0) built for x86_64-linux
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Copyright 1987-2014, Larry Wall
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Perl may be copied only under the terms of either the Artistic License or the
    [2014-07-12 15:09:08] GNU General Public License, which may be found in the Perl 5 source kit.
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Complete documentation for Perl, including FAQ lists, should be found on
    [2014-07-12 15:09:08] this system using "man perl" or "perldoc perl".  If you have access to the
    [2014-07-12 15:09:08] Internet, point your browser at http://www.perl.org/, the Perl Home Page.

# DESCRIPTION

[custom-output](https://github.com/shoichikaji/custom-output/blob/master/bin/custom-output) script allows you to exec a command and customize its output.

Here 'customize' means:

- append time, eg: 2014-07-12 15:09:08
- append hostname, eg: www.example.com
- append arbitrary strings, eg: task\_name
- write output to logfile too
- restrict output by regular expressions

## MOTIVATION

When I run a command which will take a while and whose output is messy,
it is difficult to find out which output is important
or how many steps the command is executed.

This script may solve it.
Let's say your command outputs 'success step NAME' or 'failed step NAME' every end of step.
If you exec your command by

    > custom-output --grep 'success|failed' --logfile my.log -- your-command arg1 arg2

then while the whole output is in `my.log`, STDOUT/STDERR output restrict lines
that matches `/success|failed/`.

# LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Shoichi Kaji <skaji@cpan.org>


Changes
Revision history for Perl extension App-CustomOutput

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'App-CustomOutput',
    module_name     => 'App::CustomOutput',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



lib/App/CustomOutput.pm
package App::CustomOutput;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";



1;
__END__

=encoding utf-8

=head1 NAME

App::CustomOutput - exec command with custom output

=head1 INSTALL

    > curl -LO https://raw.githubusercontent.com/shoichikaji/custom-output/master/bin/custom-output
    > chmod +x custom-output

=head1 SYNOPSIS

    > custom-output --append-time -- perl -v
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] This is perl 5, version 20, subversion 0 (v5.20.0) built for x86_64-linux
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Copyright 1987-2014, Larry Wall
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Perl may be copied only under the terms of either the Artistic License or the
    [2014-07-12 15:09:08] GNU General Public License, which may be found in the Perl 5 source kit.
    [2014-07-12 15:09:08]
    [2014-07-12 15:09:08] Complete documentation for Perl, including FAQ lists, should be found on
    [2014-07-12 15:09:08] this system using "man perl" or "perldoc perl".  If you have access to the
    [2014-07-12 15:09:08] Internet, point your browser at http://www.perl.org/, the Perl Home Page.

=head1 DESCRIPTION

L<custom-output|https://github.com/shoichikaji/custom-output/blob/master/bin/custom-output> script allows you to exec a command and customize its output.

Here 'customize' means:

=over 4

=item append time, eg: 2014-07-12 15:09:08

=item append hostname, eg: www.example.com

=item append arbitrary strings, eg: task_name

=item write output to logfile too

=item restrict output by regular expressions

=back

=head2 MOTIVATION

When I run a command which will take a while and whose output is messy,
it is difficult to find out which output is important
or how many steps the command is executed.

This script may solve it.
Let's say your command outputs 'success step NAME' or 'failed step NAME' every end of step.
If you exec your command by

    > custom-output --grep 'success|failed' --logfile my.log -- your-command arg1 arg2

then while the whole output is in C<my.log>, STDOUT/STDERR output restrict lines
that matches C</success|failed/>.

=head1 LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=cut

